packages/grafana-ui/src/components/Logs/LogRows.tsx

```diff
@@ -0,0 +1,143 @@
+import React, { PureComponent } from 'react';
+import { cx } from 'emotion';
+import { LogsModel, TimeZone, LogsDedupStrategy, LogRowModel } from '@grafana/data';
+
+import { LogRow } from './LogRow';
+import { Themeable } from '../../types/theme';
+import { withTheme } from '../../themes/index';
+import { getLogRowStyles } from './getLogRowStyles';
+
+const PREVIEW_LIMIT = 100;
+const RENDER_LIMIT = 500;
+
+export interface Props extends Themeable {
+  data: LogsModel;
+  dedupStrategy: LogsDedupStrategy;
+  highlighterExpressions: string[];
+  showTime: boolean;
+  showLabels: boolean;
+  timeZone: TimeZone;
+  deduplicatedData?: LogsModel;
+  rowLimit?: number;
+  onClickLabel?: (label: string, value: string) => void;
+  getRowContext?: (row: LogRowModel, options?: any) => Promise<any>;
+}
+
+interface State {
+  deferLogs: boolean;
+  renderAll: boolean;
+}
+
+class UnThemedLogRows extends PureComponent<Props, State> {
+  deferLogsTimer: NodeJS.Timer | null = null;
+  renderAllTimer: NodeJS.Timer | null = null;
+
+  state: State = {
+    deferLogs: true,
+    renderAll: false,
+  };
+
+  componentDidMount() {
+    // Staged rendering
+    if (this.state.deferLogs) {
+      const { data } = this.props;
+      const rowCount = data && data.rows ? data.rows.length : 0;
+      // Render all right away if not too far over the limit
+      const renderAll = rowCount <= PREVIEW_LIMIT * 2;
+      this.deferLogsTimer = setTimeout(() => this.setState({ deferLogs: false, renderAll }), rowCount);
+    }
+  }
+
+  componentDidUpdate(prevProps: Props, prevState: State) {
+    // Staged rendering
+    if (prevState.deferLogs && !this.state.deferLogs && !this.state.renderAll) {
+      this.renderAllTimer = setTimeout(() => this.setState({ renderAll: true }), 2000);
```

Comment:
It does not really matter... but rather than:
```
deferLogsTimer: NodeJS.Timer
```
can we do:
```
renderAllTimer: numer | null = null;
...
this.renderAllTimer = window.setTimeout(...
```
The NodeJS namespace seems wrong

Reply:
Good point, will change

Reply:
Yes, I've bumped into this problem several time, too.